sdk/provisioning/Azure.Provisioning/src/BicepValueOfT.cs

```diff
@@ -0,0 +1,151 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.ComponentModel;
+using Azure.Provisioning.Expressions;
+using Azure.Provisioning.Primitives;
+
+namespace Azure.Provisioning;
+
+// TODO: See the comment on BicepValue about how this will be reshaped into
+// more strongly typed models.
+
+/// <summary>
+/// Represents a value or expression.
+/// </summary>
+/// <typeparam name="T">The type of the value.</typeparam>
+public class BicepValue<T> : BicepValue
+{
+    /// <summary>
+    /// Gets or sets the literal value.  You can also rely on implicit
+    /// conversions most of the time.
+    /// </summary>
+    public T? Value
+    {
+        get
+        {
+            // If someone is referencing an output of a named resource, rather
+            // than just return null, we'll return a reference expression
+            if (IsOutput &&
+                Kind == BicepValueKind.Unset &&
+                typeof(ProvisioningConstruct).IsAssignableFrom(typeof(T)) &&
+                Self?.Construct is NamedProvisioningConstruct)
+            {
+                // TODO: This is obviously a hack but we need to decide if we'd
+                // rather have a separate type for outputs (kind of like we
+                // do with ResourceReference<T>) or whether we want to merge
+                // this concept into BicepValue<T>.
+                T val = Activator.CreateInstance<T>();
+                (val as ProvisioningConstruct)!.OverrideWithExpression(Self.GetReference());
+                return val;
+            }
```

Comment:
There's admittedly a decent amount of exploratory cruft when you run through all this that's not high enough quality for GA.  One of the benefits of generating everything as you design the API is you're free to swing for the fences because it only takes an hour to completely reshape everything.  I want folks to play with the API to get an idea of what works and what doesn't and then we'll clean this up into something a little more even keeled.